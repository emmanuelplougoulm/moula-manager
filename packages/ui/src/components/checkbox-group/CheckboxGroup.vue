<template>
  <InputGroupContainer
    :label="label"
    :error="errorText"
    :tooltip-title="tooltipTitle"
    :tooltip-description="tooltipDescription"
    :selected="selected"
  >
    <div v-for="option in options" :key="option.id">
      <input
        :id="option.value"
        v-model="selectedOptions"
        type="checkbox"
        :value="option.value"
        :name="name"
        :data-testid="`input-${option.id}`"
        :disabled="option.disabled"
        @change="onChange"
      />
      <label class="text-big-regular" :for="option.value">
        <div class="icon-container">
          <Check class="icon" />
        </div>
        {{ option.text }}
      </label>
    </div>
  </InputGroupContainer>
</template>

<script setup lang="ts">
import { ref, computed } from 'vue';
import InputGroupContainer from '../input-group-container/InputGroupContainer.vue';
import Check from '../icons/Check.vue';
import type { InputOptionType } from '../../types';

const props = defineProps({
  options: {
    type: Array as () => InputOptionType[],
    required: true
  },
  name: {
    type: String,
    required: true
  },
  label: {
    type: String,
    required: true
  },
  error: {
    type: String,
    default: ''
  },
  tooltipTitle: {
    type: String,
    default: undefined
  },
  tooltipDescription: {
    type: String,
    default: undefined
  }
});

const selectedOptions = ref([]);
const selected = computed(() => selectedOptions.value.length > 0);
const errorText = ref(props.error);

const emit = defineEmits(['select:options']);

const onChange = () => {
  errorText.value = !selected.value ? 'Required' : '';
  emit('select:options', selectedOptions.value);
};
</script>

<style scoped>
input {
  display: none;
}

label {
  display: flex;
  align-items: center;
  color: var(--neutral-20);
  background-color: var(--neutral-100);
  cursor: pointer;
  padding: 0.8125rem 1rem;
  gap: 0.75rem;
  line-height: 1.375rem;
  outline: 0.063rem solid var(--neutral-60);
  border-radius: 0.25rem;
}

label:hover {
  background-color: var(--neutral-80);
}

input:checked + label {
  outline: 0.125rem solid var(--neutral-20);
  border-radius: 0.125rem;
  background-color: var(--neutral-80);
}

input:disabled + label {
  opacity: 0.3;
}

.icon-container {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 1.125rem;
  height: 1.125rem;
  border-radius: 0.125rem;
  transition: all 0.2s;
  border: 0.125rem solid var(--neutral-40);
  background-color: var(--neutral-100);
  box-sizing: border-box;
}

input:checked + label > .icon-container {
  background-color: var(--neutral-40);
}

.icon {
  color: var(--neutral-100);
  width: 1.25rem;
  height: 1.25rem;
}
</style>
