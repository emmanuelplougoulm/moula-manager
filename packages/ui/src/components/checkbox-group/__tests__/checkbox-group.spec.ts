import { mount } from '@vue/test-utils';
import Icons from '../../icons/index';
import CheckboxGroup from '../CheckboxGroup.vue';

const inputTestId = "[data-testid='input-1']";

const options = [
  {
    id: 0,
    text: 'Option 1',
    value: 'option-1',
    disabled: false
  },
  {
    id: 1,
    text: 'Option 2',
    value: 'option-2',
    disabled: false
  }
];

const disabledOptions = [
  {
    id: 0,
    text: 'Option 1',
    value: 'option-1',
    disabled: true
  },
  {
    id: 1,
    text: 'Option 2',
    value: 'option-2',
    disabled: true
  }
];

describe('CheckboxGroup tests', () => {
  it('should display a highlighted label and no error message by default', async () => {
    const wrapper = mount(CheckboxGroup, {
      global: {
        plugins: [Icons]
      },
      props: {
        label: 'Text',
        name: 'name',
        options: options
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display a non highlighted label and no error message if an option is checked', async () => {
    const wrapper = mount(CheckboxGroup, {
      global: {
        plugins: [Icons]
      },
      props: {
        label: 'Text',
        name: 'name',
        error: 'This field is required',
        options: options
      }
    });
    const input = wrapper.find(inputTestId);
    await input.setValue(true);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display a highlighted label and an error message if an option is checked then unchecked', async () => {
    const wrapper = mount(CheckboxGroup, {
      global: {
        plugins: [Icons]
      },
      props: {
        label: 'Text',
        name: 'name',
        error: 'This field is required',
        options: options
      }
    });
    const input = wrapper.find(inputTestId);
    await input.setValue(true);
    await input.setValue(false);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display the component with disabled options', () => {
    const wrapper = mount(CheckboxGroup, {
      global: {
        plugins: [Icons]
      },
      props: {
        label: 'Text',
        options: disabledOptions,
        name: 'name'
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should emit an event when a box is checked', async () => {
    const wrapper = mount(CheckboxGroup, {
      props: {
        label: 'Text',
        options: options,
        name: 'name'
      }
    });
    const input = wrapper.find(inputTestId);
    await input.setValue(true);
    expect(wrapper.emitted()).toHaveProperty('select:options');
  });
});
