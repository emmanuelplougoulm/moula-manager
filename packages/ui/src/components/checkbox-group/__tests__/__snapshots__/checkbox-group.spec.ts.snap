// Vitest Snapshot v1

exports[`CheckboxGroup tests > should display a highlighted label and an error message if an option is checked then unchecked 1`] = `
"<div data-v-35f5b373=\\"\\" data-v-b44588ef=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-35f5b373=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default label--highlighted\\" data-testid=\\"label-item\\">Text <!--v-if--></span>
  <div data-v-b44588ef=\\"\\"><input data-v-b44588ef=\\"\\" id=\\"option-1\\" type=\\"checkbox\\" name=\\"name\\" data-testid=\\"input-0\\" value=\\"option-1\\"><label data-v-b44588ef=\\"\\" class=\\"text-big-regular\\" for=\\"option-1\\">
      <div data-v-b44588ef=\\"\\" class=\\"icon-container\\">
        <div data-v-b9dd0015=\\"\\" data-v-b44588ef=\\"\\" class=\\"container icon\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
            <path d=\\"M378 823.566 140.434 586l57.131-57.131L378 709.304l383.435-383.435L818.566 383 378 823.566Z\\"></path>
          </svg></div>
      </div> Option 1
    </label></div>
  <div data-v-b44588ef=\\"\\"><input data-v-b44588ef=\\"\\" id=\\"option-2\\" type=\\"checkbox\\" name=\\"name\\" data-testid=\\"input-1\\" value=\\"option-2\\"><label data-v-b44588ef=\\"\\" class=\\"text-big-regular\\" for=\\"option-2\\">
      <div data-v-b44588ef=\\"\\" class=\\"icon-container\\">
        <div data-v-b9dd0015=\\"\\" data-v-b44588ef=\\"\\" class=\\"container icon\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
            <path d=\\"M378 823.566 140.434 586l57.131-57.131L378 709.304l383.435-383.435L818.566 383 378 823.566Z\\"></path>
          </svg></div>
      </div> Option 2
    </label></div><span data-v-c1d8433f=\\"\\" data-v-35f5b373=\\"\\" id=\\"error-item\\" data-testid=\\"error-item\\" class=\\"text-small-regular\\">Required</span>
</div>"
`;

exports[`CheckboxGroup tests > should display a highlighted label and no error message by default 1`] = `
"<div data-v-35f5b373=\\"\\" data-v-b44588ef=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-35f5b373=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default label--highlighted\\" data-testid=\\"label-item\\">Text <!--v-if--></span>
  <div data-v-b44588ef=\\"\\"><input data-v-b44588ef=\\"\\" id=\\"option-1\\" type=\\"checkbox\\" name=\\"name\\" data-testid=\\"input-0\\" value=\\"option-1\\"><label data-v-b44588ef=\\"\\" class=\\"text-big-regular\\" for=\\"option-1\\">
      <div data-v-b44588ef=\\"\\" class=\\"icon-container\\">
        <div data-v-b9dd0015=\\"\\" data-v-b44588ef=\\"\\" class=\\"container icon\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
            <path d=\\"M378 823.566 140.434 586l57.131-57.131L378 709.304l383.435-383.435L818.566 383 378 823.566Z\\"></path>
          </svg></div>
      </div> Option 1
    </label></div>
  <div data-v-b44588ef=\\"\\"><input data-v-b44588ef=\\"\\" id=\\"option-2\\" type=\\"checkbox\\" name=\\"name\\" data-testid=\\"input-1\\" value=\\"option-2\\"><label data-v-b44588ef=\\"\\" class=\\"text-big-regular\\" for=\\"option-2\\">
      <div data-v-b44588ef=\\"\\" class=\\"icon-container\\">
        <div data-v-b9dd0015=\\"\\" data-v-b44588ef=\\"\\" class=\\"container icon\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
            <path d=\\"M378 823.566 140.434 586l57.131-57.131L378 709.304l383.435-383.435L818.566 383 378 823.566Z\\"></path>
          </svg></div>
      </div> Option 2
    </label></div>
  <!--v-if-->
</div>"
`;

exports[`CheckboxGroup tests > should display a non highlighted label and no error message if an option is checked 1`] = `
"<div data-v-35f5b373=\\"\\" data-v-b44588ef=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-35f5b373=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default\\" data-testid=\\"label-item\\">Text <!--v-if--></span>
  <div data-v-b44588ef=\\"\\"><input data-v-b44588ef=\\"\\" id=\\"option-1\\" type=\\"checkbox\\" name=\\"name\\" data-testid=\\"input-0\\" value=\\"option-1\\"><label data-v-b44588ef=\\"\\" class=\\"text-big-regular\\" for=\\"option-1\\">
      <div data-v-b44588ef=\\"\\" class=\\"icon-container\\">
        <div data-v-b9dd0015=\\"\\" data-v-b44588ef=\\"\\" class=\\"container icon\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
            <path d=\\"M378 823.566 140.434 586l57.131-57.131L378 709.304l383.435-383.435L818.566 383 378 823.566Z\\"></path>
          </svg></div>
      </div> Option 1
    </label></div>
  <div data-v-b44588ef=\\"\\"><input data-v-b44588ef=\\"\\" id=\\"option-2\\" type=\\"checkbox\\" name=\\"name\\" data-testid=\\"input-1\\" value=\\"option-2\\"><label data-v-b44588ef=\\"\\" class=\\"text-big-regular\\" for=\\"option-2\\">
      <div data-v-b44588ef=\\"\\" class=\\"icon-container\\">
        <div data-v-b9dd0015=\\"\\" data-v-b44588ef=\\"\\" class=\\"container icon\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
            <path d=\\"M378 823.566 140.434 586l57.131-57.131L378 709.304l383.435-383.435L818.566 383 378 823.566Z\\"></path>
          </svg></div>
      </div> Option 2
    </label></div>
  <!--v-if-->
</div>"
`;

exports[`CheckboxGroup tests > should display the component with disabled options 1`] = `
"<div data-v-35f5b373=\\"\\" data-v-b44588ef=\\"\\" class=\\"container\\"><span data-v-0704e93f=\\"\\" data-v-35f5b373=\\"\\" id=\\"label-item\\" class=\\"text-small-semi-bold label--default label--highlighted\\" data-testid=\\"label-item\\">Text <!--v-if--></span>
  <div data-v-b44588ef=\\"\\"><input data-v-b44588ef=\\"\\" id=\\"option-1\\" type=\\"checkbox\\" name=\\"name\\" data-testid=\\"input-0\\" disabled=\\"\\" value=\\"option-1\\"><label data-v-b44588ef=\\"\\" class=\\"text-big-regular\\" for=\\"option-1\\">
      <div data-v-b44588ef=\\"\\" class=\\"icon-container\\">
        <div data-v-b9dd0015=\\"\\" data-v-b44588ef=\\"\\" class=\\"container icon\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
            <path d=\\"M378 823.566 140.434 586l57.131-57.131L378 709.304l383.435-383.435L818.566 383 378 823.566Z\\"></path>
          </svg></div>
      </div> Option 1
    </label></div>
  <div data-v-b44588ef=\\"\\"><input data-v-b44588ef=\\"\\" id=\\"option-2\\" type=\\"checkbox\\" name=\\"name\\" data-testid=\\"input-1\\" disabled=\\"\\" value=\\"option-2\\"><label data-v-b44588ef=\\"\\" class=\\"text-big-regular\\" for=\\"option-2\\">
      <div data-v-b44588ef=\\"\\" class=\\"icon-container\\">
        <div data-v-b9dd0015=\\"\\" data-v-b44588ef=\\"\\" class=\\"container icon\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
            <path d=\\"M378 823.566 140.434 586l57.131-57.131L378 709.304l383.435-383.435L818.566 383 378 823.566Z\\"></path>
          </svg></div>
      </div> Option 2
    </label></div>
  <!--v-if-->
</div>"
`;
